src/libcore/lazy.rs

```diff
@@ -0,0 +1,379 @@
+//! Lazy values and one-time initialization of static data.
+
+use crate::cell::{Cell, UnsafeCell};
+use crate::fmt;
+use crate::mem;
+use crate::ops::Deref;
+
+/// A cell which can be written to only once.
+///
+/// Unlike `RefCell`, a `OnceCell` only provides shared `&T` references to its value.
+/// Unlike `Cell`, a `OnceCell` doesn't require copying or replacing the value to access it.
+///
+/// # Examples
+///
+/// ```
+/// #![feature(once_cell)]
+///
+/// use std::lazy::OnceCell;
+///
+/// let cell = OnceCell::new();
+/// assert!(cell.get().is_none());
+///
+/// let value: &String = cell.get_or_init(|| {
+///     "Hello, World!".to_string()
+/// });
+/// assert_eq!(value, "Hello, World!");
+/// assert!(cell.get().is_some());
+/// ```
+#[unstable(feature = "once_cell", issue = "74465")]
+pub struct OnceCell<T> {
+    // Invariant: written to at most once.
+    inner: UnsafeCell<Option<T>>,
+}
```

Comment:
Do we rely on implicit `!Sync` from `UnsafeCell`, or do we try to be explicit? cc @RalfJung 

(Sorry if this was asked already, I just saw this PR and thought it was neat, and wasn't sure if this `OnceCell` was meant to be thread-safe or not)

Reply:
Ah the `OnceCell` here in `core` isn’t meant to be thread-safe, so we’re just relying on `UnsafeCell`. I’m actually not sure whether we’ve got a preferred approach for being explicit about clobbering auto-traits... I know it’s bitten us sometimes.

Reply:
It didn't help that GitHub hides the `std::sync` version by default, it was clearer once I found it.

Reply:
I'm personally a big fan of making things explicit.